In [ ]:
# %pip install pandas openpyxl 
#   use of excel writer

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

import pingouin as pg

In [ ]:
#   File Beer Profiles and ratings.csv
#   --------------------------------------

# This file contains 1 Tabs  with Beer dateset detailled profile and ratings info
    #   Objective of this script - check relations between dimension concepts


In [2]:
# Fetch raw data

rawdata = pd.read_csv(r'C:\Users\t.vincotte\Documents\CODE\TFE\EoF - Work\Beer_Profile\Source_Files\beer_reviews.csv') 
 

In [3]:
rawdata.count()

brewery_id            1586614
brewery_name          1586599
review_time           1586614
review_overall        1586614
review_aroma          1586614
review_appearance     1586614
review_profilename    1586266
beer_style            1586614
review_palate         1586614
review_taste          1586614
beer_name             1586614
beer_abv              1518829
beer_beerid           1586614
dtype: int64

In [4]:
#rawdata.info()
str = rawdata.info()
structure = pd.DataFrame(str)  #  ne met pas en dataframe
structure
# structure.to_excel("Beer_datasets_structures",sheet_name='Beer data set details') # n ecrit pas le rawdata.info dans le fichier

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   brewery_id          1586614 non-null  int64  
 1   brewery_name        1586599 non-null  object 
 2   review_time         1586614 non-null  int64  
 3   review_overall      1586614 non-null  float64
 4   review_aroma        1586614 non-null  float64
 5   review_appearance   1586614 non-null  float64
 6   review_profilename  1586266 non-null  object 
 7   beer_style          1586614 non-null  object 
 8   review_palate       1586614 non-null  float64
 9   review_taste        1586614 non-null  float64
 10  beer_name           1586614 non-null  object 
 11  beer_abv            1518829 non-null  float64
 12  beer_beerid         1586614 non-null  int64  
dtypes: float64(6), int64(3), object(4)
memory usage: 157.4+ MB


Empty DataFrame
Columns: []
Index: []

In [5]:
rawdata.head(1)

brewery_id     brewery_name  review_time  review_overall  review_aroma  \
0       10325  Vecchio Birraio   1234817823             1.5           2.0   

   review_appearance review_profilename  beer_style  review_palate  \
0                2.5            stcules  Hefeweizen            1.5   

   review_taste     beer_name  beer_abv  beer_beerid  
0           1.5  Sausa Weizen       5.0        47986

In [ ]:
#   Only show columns we want to see - methods not used here
#   ----------------------------------------------------------
        # Select the ones you want
            #   df1 = df[['a','d']]


Beer dimension check on whole set >> too large >> grouping needed
-------------------------------------------------------------------

 Sheet is too large! Your sheet size is: 1586614, 6 Max sheet size is: 1048576, 16384 >>> other approach to be taken
 For dimension verification, grouping on potebtial dimensions is enough

 Grouping documentation :    https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/

            import pandas as pd
            df = pd.read_csv("https://media.geeksforgeeks.org/wp-content/uploads/nba.csv")

            grouping = df.groupby(['Team', 'Position'])
            print(grouping.first())

In [ ]:
#   Beer dimension check - Grouping counts() evaluation
#   ----------------------------------------------------------

    #   https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/

df_beer_count = rawdata[['beer_beerid','beer_name','brewery_id','brewery_name','beer_style','review_profilename']].groupby(['beer_name']).count()
display(df_beer_count)

df_style_count = rawdata[['beer_style','beer_name']].groupby(['beer_style']).count()
display(df_style_count)

df_brewery_count = rawdata[['brewery_id','brewery_name']].groupby(['brewery_name','brewery_id']).count()
display(df_brewery_count)

df_reviewer_count = rawdata[['review_profilename','beer_name']].groupby(['review_profilename']).count()
display(df_reviewer_count)

df_dimensions_without_time_count = rawdata[['beer_beerid','beer_name','brewery_id','brewery_name','beer_style','review_profilename']].groupby(['beer_name','brewery_name','beer_style','review_profilename']).count()
display(df_dimensions_without_time_count)

df_dimensions_without_reviewers = rawdata[['beer_name','brewery_id','brewery_name','beer_style']].groupby(['beer_name','brewery_name','beer_style']).count()
display(df_dimensions_without_reviewers)



Outcome of grouping evaluations
--------------------------------
>   66055 beers
>   105 beer styles
>   5838 breweries
>   33387 reviewers
>   Combination of unique beer, reviewer : 1571324 rows vs a total of 1586614 entries - business, a few professionals, many occasional
>   For dimension check, grouping all dimensions except reviewers is fine >> df_dimensions_without_reviewers 65743 combinations

df_dimensions_without_reviewers will be used to generate the sheets diemnsion verification

In [ ]:
#   Reorder DataFrame columns : https://www.geeksforgeeks.org/change-the-order-of-a-pandas-dataframe-columns-in-python/ 
#       df_reordered = df.loc[:, ['Maths Score', 'Name', 'Sr.no']]

        #   REORDER COLUMNS TO REVISIT
                #   df1_by_beer = df1
                #   df1_by_beer

        # had issue to reorder a dataframe other one

In [64]:
#   Beer dimension check - rough listing - create spreadsheet 
#   ----------------------------------------------------------

    #   https://www.geeksforgeeks.org/exporting-a-pandas-dataframe-to-an-excel-file/
    #   https://stackoverflow.com/questions/42370977/how-to-save-a-new-sheet-in-an-existing-excel-file-using-pandas
    #   https://saturncloud.io/blog/how-to-add-a-worksheet-to-an-existing-excel-file-with-pandas/ 


df1_source = rawdata[['beer_name','brewery_id','brewery_name','beer_style','beer_abv']].drop_duplicates()
df1 = df1_source

df1.to_excel("EDA_DimCheck_BeerReview.xlsx",sheet_name='rough_listing')

        #   Note : (Re)creates spreadsheet as running - below - use method to append sheet

In [65]:
#   Beer dimension check - List by style
#   ---------------------------------------
    #   Use method to append sheet to existing excel file

df1_by_style = df1[['beer_style','beer_name','brewery_name','beer_abv']].sort_values(by=['beer_name']).sort_values(by=['beer_style'])
with pd.ExcelWriter('EDA_DimCheck_BeerReview.xlsx', engine='openpyxl', mode='a') as writer:
    # Write the new DataFrame to a new sheet
    df1_by_style.to_excel(writer, sheet_name='list_by_style', index=True)

# df1_by_style.to_excel("EDA_DimCheck_BeerDataset.xlsx",sheet_name='list_by_style')


In [66]:
#   Beer dimension check - List by brewery
#   ---------------------------------------
   #   Use method to append sheet to existing excel file

df1_by_brewery = df1[['brewery_name','beer_name','beer_style','beer_abv']].sort_values(by=['beer_name']).sort_values(by=['brewery_name']) # type: ignore
with pd.ExcelWriter('EDA_DimCheck_BeerReview.xlsx', engine='openpyxl', mode='a') as writer:
    # Write the new DataFrame to a new sheet
    df1_by_brewery.to_excel(writer, sheet_name='list_by_brewery', index=True)


In [67]:
#   Beer dimension check - List by style_brewery
#   ---------------------------------------------
   #   Use method to append sheet to existing excel file

df1_by_brewery_style = df1[['beer_style','brewery_name','beer_name','beer_abv']].sort_values(by=['beer_name']).sort_values(by=['brewery_name']).sort_values(by=['beer_style'])
with pd.ExcelWriter('EDA_DimCheck_BeerReview.xlsx', engine='openpyxl', mode='a') as writer:
    # Write the new DataFrame to a new sheet
    df1_by_brewery_style.to_excel(writer, sheet_name='list_by_brewery_style', index=True)


In [68]:
#   Beer dimension check - List by brewery_style
#   ---------------------------------------------
   #   Use method to append sheet to existing excel file

df1_by_brewery_style = df1[['brewery_name','beer_style','beer_name','beer_abv']].sort_values(by=['beer_name']).sort_values(by=['beer_style']).sort_values(by=['brewery_name'])
with pd.ExcelWriter('EDA_DimCheck_BeerReview.xlsx', engine='openpyxl', mode='a') as writer:
    # Write the new DataFrame to a new sheet
    df1_by_brewery_style.to_excel(writer, sheet_name='list_by_style_brewery', index=True)


In [71]:
rawdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1586614 entries, 0 to 1586613
Data columns (total 13 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   brewery_id          1586614 non-null  int64  
 1   brewery_name        1586599 non-null  object 
 2   review_time         1586614 non-null  int64  
 3   review_overall      1586614 non-null  float64
 4   review_aroma        1586614 non-null  float64
 5   review_appearance   1586614 non-null  float64
 6   review_profilename  1586266 non-null  object 
 7   beer_style          1586614 non-null  object 
 8   review_palate       1586614 non-null  float64
 9   review_taste        1586614 non-null  float64
 10  beer_name           1586614 non-null  object 
 11  beer_abv            1518829 non-null  float64
 12  beer_beerid         1586614 non-null  int64  
dtypes: float64(6), int64(3), object(4)
memory usage: 157.4+ MB


In [ ]:
#   COUNT DIFFERENTS
#   ------------------------------

# Top : 1.586.614 total row counts
#   Rows where count are different - might be NULL



In [ ]:
#     Check if dupplicate beer evaluations
#   -------------------------------------
#       #   https://sparkbyexamples.com/pandas/pandas-filter-rows-by-conditions/#:~:text=You%20can%20filter%20the%20rows,()%2C%20or%20apply()%20function.
        #   

df_eval_before_count = rawdata[['beer_beerid','beer_abv','beer_name','brewery_id','brewery_name','beer_style','review_profilename','review_time']].count()
#display(df_eval_before_count)

df_duplicated = rawdata[['beer_beerid','beer_abv','beer_name','brewery_id','brewery_name','beer_style','review_profilename','review_time']].duplicated()
duplicated_sorted = pd.DataFrame(df_duplicated).sort_values(by=0,ascending=False)
        # duplicates = duplicated_sorted.loc[duplicated_sorted[:]].isin(True)
display(duplicates)


df_dropped_duplicates = rawdata[['beer_beerid','beer_abv','beer_name','brewery_id','brewery_name','beer_style','review_profilename','review_time']].drop_duplicates()


df_eval_after_count = df_dropped_duplicates.count()
#display(df_eval_after_count)


Sounds quirte cmlose -- NEED TO UNDERSTAND THE SMALL DIFFERENCES IN THE COUNT  ??
WAY TO SHOW DUPPLICATE LINES ??
>>> FEW DUPLICATE ROWS .... Might be some NULLs

>>> Make an heatmap schema - search rows with NULL and treat

In [14]:
#     Check if dupplicate beer names
#   -------------------------------------
#   
 
df2x = df1[['Name']].drop_duplicates
#df2 = pd.DataFrame[df2x]
df2x

#df2_duplicate_beer_check = df2.count()
#df2_duplicate_beer_check 

<bound method DataFrame.drop_duplicates of                                           Name
0                                        Amber
1                                   Double Bag
2                               Long Trail Ale
3                                 Doppelsticke
4                 Sleigh'r Dark Doüble Alt Ale
...                                        ...
3192                           Winter Shredder
3193                        The First Snow Ale
3194                       Red Nose Winter Ale
3195                      Fish Tale Winterfish
3196  Frosted Frog Christmas Ale - Barrel-Aged

[3197 rows x 1 columns]>

In [94]:
rawdata.dtypes

brewery_id              int64
brewery_name           object
review_time             int64
review_overall        float64
review_aroma          float64
review_appearance     float64
review_profilename     object
beer_style             object
review_palate         float64
review_taste          float64
beer_name              object
beer_abv              float64
beer_beerid             int64
dtype: object

Afre drop dupplicate still amount of rows >> no duplicate in beer name

3197 records